src/kernel/emulation/linux/misc/sysctl_machdep.c

```diff
@@ -51,3 +77,132 @@ sysctl_handler(handle_brand_string)
 	return 0;
 }
 
+sysctl_handler(handle_max_basic)
+{
+	setup(0);
+
+
+    __cpuid(level, eax, ebx, ecx, edx);
+
+
+	sprintf(old,"%d",eax);
+
+	return 0;
+}
+
+sysctl_handler(handle_family)
+{
+    setup(1);
+
+    __cpuid(level, eax, ebx, ecx, edx);
+
+    eax = eax >> 7;
+    eax &= 15;
+
+    sprintf(old,"%d",eax);
+
+    return 0;
+}
+
+sysctl_handler(handle_model)
+{
+    setup(1);
+
+    __cpuid(level, eax, ebx, ecx, edx);
+
+    eax = eax >> 3;
+    eax &= 15;
+
+    sprintf(old,"%d",eax);
+
+    return 0;
+}
+
+sysctl_handler(handle_stepping)
+{
+    setup(1);
+
+    __cpuid(level, eax, ebx, ecx, edx);
+
+    eax &= 15;
+
+    sprintf(old,"%d",eax);
+
+    return 0;
+}
+
+sysctl_handler(handle_brand_string)
+{
+    unsigned int level = 0;
+    unsigned int eax = 0x80000000;
+
+    unsigned int test[3];
+
+    __cpuid(level,eax,test, test+1,test+2);
+
+    if(eax < 0x80000004) // the information is not implemented
+        return 2;
+
+    eax = 0x80000001;
+
+    union
+    {
+        unsigned int brand[12];
+        char name[49];
+    } v;
+
+    __cpuid(level, eax, v.brand+0x1, v.brand+0x2, v.brand+0x3);
+    v.brand[0x0] = eax;
+    eax = 0x80000002;
+
+    __cpuid(level, eax, v.brand+0x5, v.brand+0x6, v.brand+0x7);
+    v.brand[0x4] = eax;
+    eax = 0x80000003;
+
+    __cpuid(level, eax, v.brand+0x9, v.brand+0xa, v.brand+0xb);
+    v.brand[0x8] = eax;
+
+    v.name[48] = 0;
+    copyout_string(v.name, (char*) old, oldlen);
+
+    return 0;
+}
+sysctl_handler(handle_features)
+{
+    setup(1);
+
+    char *features[] = {"FPU","VME", "DE", "PSE", "TSC", "MSR", "PAE", "MCE", "CX8", "APIC", NULL,"SEP","PGE","MCA",
+                        "CMOV", "PAT", "PSE-36", "PSN", "CLFSH", "DS", "ACPI", "MMX", "FXSR", "SSE", "SSE2", "SS",
+                        "HTT", "TM", NULL, "PBE"};
```

Comment:
The second `NULL` is actually in the wrong place (it's at index 30, but it *should* be at index 20, according to [Wikipedia](https://en.wikipedia.org/wiki/CPUID#EAX=1:_Processor_Info_and_Feature_Bits)); plus, there's 31 elements, not 32.
```suggestion
                        "CMOV", "PAT", "PSE-36", "PSN", "CLFSH", NULL, "DS", "ACPI", "MMX", "FXSR", "SSE", "SSE2", "SS",
                        "HTT", "TM", "IA64", "PBE"};
```

Reply:
I made it according to [this site](https://c9x.me/x86/html/file_module_x86_id_45.html)  and there the 30 is the reserved one. The 32 elements are good because the counting starts with 0 and ends with 31.

Reply:
Oh, now I see. I had 31 elements, not 32.